mage_ai/io/google_sheets.py

```diff
@@ -0,0 +1,289 @@
+
+
+class GoogleSheets(BaseFile):
+    """
+    Handles data transfer between Google Sheet and the Mage app.
+    Uses the gspread library to handle IO. Supports loading sheets from any one of:
+    - Sheet ID
+    - Sheet URL
+    - Sheet Name
+
+    And worksheets using any one of:
+    - Worksheet Name
+    - Worksheet Position
+
+    Exporting worksheets currently requires an existing worksheet to be present.
+
+    If GOOGLE_APPLICATION_CREDENTIALS environment is set, no further arguments are needed
+    other than those specified below. Otherwise, use the factory method `with_config` to construct
+    the data loader using manually specified credentials.
+    """
+
+    def __init__(
+        self,
+        **kwargs,
+    ) -> None:
+        """
+        Initializes data loader from a Google Sheet. This requires the Sheets & Drive APIs
+        to be enabled.
+
+        To authenticate (and authorize) access to Google Sheets, credentials must be
+        provided.
+
+        Below are the different ways to access those credentials:
+        - Define the `GOOGLE_APPLICATION_CREDENTIALS` environment variable to point to a
+          service account key. In this case no other no other parameters need to be
+          specified.
+        - Manually pass in the `google.oauth2.service_account.Credentials` object with the
+        keyword argument `credentials`
+        - Manually pass in the path to the credentials with the keyword argument
+        `path_to_credentials`.
+        - Manually define the service key key-value set to use (such as a dictionary containing
+        the same parameters as a service key) with the keyword argument `credentials_mapping`
+
+        All keyword arguments except for `path_to_credentials` and `credentials_mapping` will be
+        passed to the Google Sheet client, accepting all other configuration settings there.
+        """
+
+        super().__init__(verbose=kwargs.get('verbose', True))
+
+        if kwargs.get('verbose') is not None:
+            kwargs.pop('verbose')
+
+        scopes = [
+            'https://www.googleapis.com/auth/spreadsheets',
+            'https://www.googleapis.com/auth/drive',
+        ]
+
+        credentials = kwargs.get('credentials')
+        if credentials is None:
+            if 'credentials_mapping' in kwargs:
+                mapping_obj = kwargs.pop('credentials_mapping')
+                if mapping_obj is not None:
+                    credentials = service_account.Credentials.from_service_account_info(
+                        mapping_obj, scopes=scopes
+                    )
+            if 'path_to_credentials' in kwargs:
+                path = kwargs.pop('path_to_credentials')
+                if path is not None:
+                    credentials = service_account.Credentials.from_service_account_file(
+                        path, scopes=scopes
+                    )
+            if 'credentials' in kwargs:
+                kwargs.pop('credentials')
```

Comment:
I think this should be outside of the `if credentials is None:` block

Reply:
I took this directly from the BigQuery code [here](https://github.com/mage-ai/mage-ai/blob/f07ba491bd048eceadc1010ed97dd9841f25220b/mage_ai/io/bigquery.py#L52), so I _think_ this is correct.

Reply:
ok i guess i could see how that would make sense too